_posts/2021-04-05-maven-bom.md

```diff
@@ -0,0 +1,312 @@
+
+A `dependencyManagement`  section contains dependency  elements. Each dependency is a lookup reference for Maven to determine the version to select for transitive dependencies. The version of the dependency is mandatory in this section. However, outside `dependencyManagement`  section, you can omit the version from your dependencies, and Maven will select the correct version of the transitive dependencies from the list of dependencies provided in `dependencyManagement`.
+
+We should note that defining a dependency in the `dependencyManagement` section doesn't add it to the dependency tree of the project, it is used just for lookup reference. Instead, for any dependency in a project's dependency tree, if it appears in the BOM file, then the version from the BOM will be used.
+
+A better way to understand the use of `dependencyManagement` is through an example. Let's go back to our previous example with the Spring dependencies. Now we are going to play with the `spring-beans` dependency. When we executed the command `mvn dependency:tree`, the version resolved for `spring-beans` was `5.3.5`.
+
+Using `dependencyManagement` we can override this version and select the version that we want. All that we have to do is to add the following to our POM file:
+
+```xml
+<dependencyManagement>
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-beans</artifactId>
+            <version>5.2.13.RELEASE</version>
+        </dependency>
+    </dependencies>
+</dependencyManagement>
+```
+
+Now we want Maven to resolve version `5.2.13.RELEASE` instead of `5.3.5`.
+
+Let's execute the command `mvn dependency:tree` one more time. The result is:
+
+```markdown
++- org.springframework:spring-context:jar:5.3.5:compile
+|  +- org.springframework:spring-aop:jar:5.3.5:compile
+|  +- org.springframework:spring-beans:jar:5.2.13.RELEASE:compile
+|  +- org.springframework:spring-core:jar:5.3.5:compile
+|  |  \- org.springframework:spring-jcl:jar:5.3.5:compile
+|  \- org.springframework:spring-expression:jar:5.3.5:compile
+\- org.springframework.security:spring-security-web:jar:5.4.5:compile
+   +- org.springframework.security:spring-security-core:jar:5.4.5:compile
+   \- org.springframework:spring-web:jar:5.2.13.RELEASE:compile
+```
+
+In the dependency tree, we find the `5.2.13.RELEASE` version for `spring-beans`. This is the version that we wanted Maven to resolve for each `spring-beans` transitive dependency.
+
+If `spring-beans` was a direct dependency, in order to take advantage of the BOM file, we will no longer have to set the version when adding the dependency:
+
+```xml
+<dependency>
+    <groupId>org.springframework</groupId>
+    <artifactId>spring-beans</artifactId>
+</dependency>
+```
+
+This way, Maven will resolve the version using the information provided in the BOM file.
```

Comment:
Again, we don't know what a BOM file is, yet.

Reply:
I replaced BOM with `dependencyManagement` section since we were talking about it earlier.